### PR TITLE
feat: dictionary implementation

### PR DESCRIPTION
**Pull request by malandis**
This PR:

1. Creates a separate incubating client
2. Implements the core dictionary functionality (`DictionarySet`, `DictionaryGet`, `DictionaryDelete`, `DictionaryRemoveField` and their multi counterparts).

For the dictionary implementation, we created a separate incubating client. In the future the incubating client will be in a separate repository; for now it is in a separate namespace. In preparation for this we made the incubating client only depend on a client interface. This interface defines the contract that a production client should implement. We chose this pattern as opposed to inheritance or reflection to: (1) loosely couple the incubating client to the production one, and (2) to be able to have client documentation available to both the production and incubating clients.

To limit accessibility of internal code, we refactored to a separate namespace. To share with the incubating client, we increased visibility where necessary.

In preparation for the repository split, the incubating client has its own set of integration tests.

### CODE REVIEW

**Comment by kvcache** [comment]
I'm reading this but it's quite big, thanks for your patience.
Please note for merging that you will need to adjust the commit message profoundly, as the squash of all these commit messages does not represent the change very well.

**Review by kvcache — COMMENTED** [review]
Nice start - there are a few things I noticed in here; and probably there are things I missed due to the size. More eyes would be good.

Also looks like the names in here aren't matching the proposal in The Document.

File: Momento/Incubating/Responses/CacheDictionaryGetResponse.cs [hunk 1]
```diff
@@ -1,28 +1,27 @@
...
-    public byte[]? Bytes()
+    public string? String()
     {
-        throw new NotImplementedException();
+        if (Bytes == null)
+        {
+            return null;
+        }
+        return Encoding.UTF8.GetString(Bytes);
     }
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionaryGetResponse.cs:26]: (1) Want to remind users it's doing something non-trivial effort (string encoding) 
(2) String encoding could raise an exception if the user calls `String()` on something from the cache that wasn't a string. I think it's bad form to have a property that could raise an exception

File: Momento/ISimpleCacheClient.cs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    /// <returns>CacheGetResponse containing the status of the get operation and the associated value.</returns>
+    public CacheGetResponse Get(string cacheName, byte[] key);
+
+    /// <summary>
+    /// Remove the key from the cache.
+    /// </summary>
+    /// <param name="cacheName">Name of the cache to delete the key from.</param>
+    /// <param name="key">The key to delete.</param>
+    /// <returns>Task containing the result of the delete operation.</returns>
+    public CacheDeleteResponse Delete(string cacheName, byte[] key);
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/ISimpleCacheClient.cs:164]: > this doesn't return a Task. (many of these comments refer to Task response which is not present)

Thanks. Must have had a typo.

> why do we need these non-task methods again?

I proposed getting rid of them. It's still a proposal.

File: Momento/Incubating/Responses/CacheDictionaryGetResponse.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kvcache) [reply Momento/Incubating/Responses/CacheDictionaryGetResponse.cs:26]: So call it `DecodeString` or `AsString`?

File: Momento/Incubating/Internal/ScsDataClient.cs [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+    }
+
+    public async Task<CacheDictionaryGetMultiResponse> DictionaryGetMultiAsync(string cacheName, string dictionaryName, IEnumerable<string> fields)
+    {
+        return await SendDictionaryGetMultiAsync(cacheName, dictionaryName, fields.Select(field => Convert(field)));
+    }
+
+    private async Task<CacheDictionaryGetMultiResponse> SendDictionaryGetMultiAsync(string cacheName, string dictionaryName, IEnumerable<ByteString> fields)
+    {
+        _DictionaryGetRequest request = new() { DictionaryName = Convert(dictionaryName) };
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Internal/ScsDataClient.cs:265]: It's pretty nifty. The above is equivalent to:

```csharp
        _DictionaryGetRequest request = new _DictionaryGetRequest();
        request.DictionaryName = Convert(dictionaryName);
```

File: Momento/Incubating/Responses/CacheDictionaryGetMultiResponse.cs [hunk 4]
```diff
@@ -1,34 +1,34 @@
...
-        throw new NotImplementedException();
+        get => responses.Select(response => response.Status).ToList();
     }
 
-    public List<byte[]?> ValuesAsBytes()
+    public List<string?> Strings()
     {
-        throw new NotImplementedException();
+        return responses.Select(response => response.String()).ToList();
     }
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionaryGetMultiResponse.cs:28]: > Why not return IEnumerable<string?> instead and omit ToList()?

Good call. I think I converted to list when intiially writing tests. Not necessary.

File: Momento/ISimpleCacheClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kvcache) [reply Momento/ISimpleCacheClient.cs:164]: I see green lines here - can you justify the presence of these non-task methods in the C# sdk?

File: Momento/Incubating/Responses/CacheDictionaryGetMultiResponse.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionaryGetMultiResponse.cs]: Fixed in 7c84e5a

File: Momento/Incubating/Responses/CacheDictionarySetResponse.cs [hunk 5]
```diff
@@ -1,36 +1,35 @@
...
 
-    public byte[] KeyAsBytes()
+    public byte[] FieldToByteArray()
     {
-        throw new NotImplementedException();
+        return (byte[])field;
     }
 
-    public string Key(Encoding? encoding = null)
+    public string FieldToString()
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionarySetResponse.cs:21]: https://github.com/momentohq/client-sdk-csharp/pull/88#discussion_r931537626

File: Momento/Incubating/Responses/CacheDictionaryRemoveFieldResponse.cs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+namespace MomentoSdk.Incubating.Responses;
+
+public class CacheDictionaryRemoveFieldResponse
+{
+}
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionaryRemoveFieldResponse.cs]: Fixed in 0c3a8c0

File: Momento/Incubating/SimpleCacheClient.cs [hunk 7]
```diff
@@ -1,103 +1,863 @@
...
+            throw new ArgumentNullException(nameof(dictionaryName));
+        }
+        if (field == null)
+        {
+            throw new ArgumentNullException(nameof(field));
+        }
+        if (value == null)
+        {
+            throw new ArgumentNullException(nameof(value));
+        }
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/SimpleCacheClient.cs:200]: Already considered `ArgumentNotNull` but it's not available in the version we are using :cry: .

I was not aware of the `NotNull` attribute and will check it out. Thanks.

[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/SimpleCacheClient.cs]: Fixed in ea02843

File: Momento/Internal/ScsDataClient.cs [hunk 8]
```diff
@@ -8,22 +8,64 @@
...
 
+    protected ByteString Convert(byte[] bytes)
+    {
+        return ByteString.CopyFrom(bytes);
+    }
+
+    protected ByteString Convert(string s)
+    {
+        return ByteString.CopyFromUtf8(s);
+    }
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Internal/ScsDataClient.cs]: Gone a step further and indulged myself in C# features. In 33d84b0 I implemented these as extension methods to `byte[]` and `string`.

File: Momento/Internal/DataGrpcManager.cs [hunk 9]
```diff
@@ -6,18 +6,19 @@
...
 
-internal sealed class DataGrpcManager : IDisposable
+public class DataGrpcManager : IDisposable
 {
     private readonly GrpcChannel channel;
     public Scs.ScsClient Client { get; }
 
     private readonly string version = "csharp:" + GetAssembly(typeof(MomentoSdk.Responses.CacheGetResponse)).GetName().Version.ToString();
 
-    internal DataGrpcManager(string authToken, string endpoint)
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Internal/DataGrpcManager.cs:18]: Even if it's just the domain name?

File: Momento/Incubating/Responses/CacheDictionarySetMultiResponse.cs [hunk 10]
```diff
@@ -6,20 +6,22 @@ namespace MomentoSdk.Incubating.Responses;
...
-        throw new NotImplementedException();
+        DictionaryName = dictionaryName;
+        this.items = items;
     }
-    public Dictionary<byte[], byte>? DictionaryAsBytes()
+
+    public IEnumerable<KeyValuePair<byte[], byte[]>> ItemsAsByteArrays()
     {
-        throw new NotImplementedException();
+        return (IEnumerable<KeyValuePair<byte[], byte[]>>)items;
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionarySetMultiResponse.cs]: After discussing what the set response objects should be doing, I removed this code in b3867ee.

File: Momento/Incubating/Responses/CacheDictionarySetResponse.cs [hunk 11]
```diff
@@ -1,36 +1,35 @@
...
     {
-        throw new NotImplementedException();
+        return (string)field;
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionarySetResponse.cs]: After discussing what the set response objects should be doing, I removed this code in b3867ee.

File: Momento/Internal/DataGrpcManager.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Internal/DataGrpcManager.cs:18]: Reverted in 06d2cca

File: Momento/ISimpleCacheClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/ISimpleCacheClient.cs:164]: Will revisit in next PR

File: Momento/Incubating/Responses/CacheDictionaryGetResponse.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Responses/CacheDictionaryGetResponse.cs:26]: We'll flesh out a spec for the names so they are consistent across SDKs, then sweep through these in a future PR.

File: Momento/Incubating/Internal/ScsDataClient.cs [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+
+    private CacheDictionaryGetResponse SendDictionaryGet(string cacheName, string dictionaryName, ByteString field)
+    {
+        _DictionaryGetRequest request = new() { DictionaryName = Convert(dictionaryName) };
+        request.DictionaryKeys.Add(field);
+
+        try
+        {
+            var response = this.grpcManager.Client.DictionaryGet(request, MetadataWithCache(cacheName), deadline: CalculateDeadline());
+            return new CacheDictionaryGetResponse(response.DictionaryBody[0]);
```
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/Incubating/Internal/ScsDataClient.cs:103]: It's already in a try-except block where an out-of-bounds exception would be caught, then thrown as a ClientSdkException. I'll leave a TODO for our upcoming exception handling UX revamp.

File: Momento/ISimpleCacheClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (malandis) [reply Momento/ISimpleCacheClient.cs]: Fixed the comments in 1181245

File: Momento/Incubating/Responses/CacheDictionarySetMultiResponse.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (kvcache) [reply Momento/Incubating/Responses/CacheDictionarySetMultiResponse.cs]: i love me some deleted code

File: Momento/Incubating/SimpleCacheClient.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kvcache) [reply Momento/Incubating/SimpleCacheClient.cs]: looks much nicer, and removes over 400 lines 🎉

**Review by kvcache — APPROVED** [review]
cool, I'm on board.

**Comment by cprice404** [comment]
I haven't had time to start looking at the code yet - hope to do that right after lunch.  But I am 💯 on the description of the choices about how to structure the incubating client, interface, accessors, etc.

**Comment by malandis** [comment]
> I haven't had time to start looking at the code yet - hope to do that right after lunch. But I am 100 on the description of the choices about how to structure the incubating client, interface, accessors, etc.

Thank you. I was seeking your feedback about this in particular.

File: Momento/ISimpleCacheClient.cs [hunk 13]
```diff
@@ -0,0 +1,248 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using MomentoSdk.Responses;
+namespace MomentoSdk;
+
+public interface ISimpleCacheClient : IDisposable
```
Comment (cprice404) [inline Momento/ISimpleCacheClient.cs:7]: ❤️

File: Momento/Incubating/SimpleCacheClient.cs [hunk 14]
```diff
@@ -1,103 +1,559 @@
...
 namespace MomentoSdk.Incubating;
 
-public class SimpleCacheClient : MomentoSdk.SimpleCacheClient
+public class SimpleCacheClient : ISimpleCacheClient
 {
-    public SimpleCacheClient(string authToken, uint defaultTtlSeconds) : base(authToken, defaultTtlSeconds)
+    private readonly ISimpleCacheClient simpleCacheClient;
+    private readonly ScsDataClient dataClient;
+
+    public SimpleCacheClient(ISimpleCacheClient simpleCacheClient, string authToken, uint defaultTtlSeconds, uint? dataClientOperationTimeoutMilliseconds = null)
```
Comment (cprice404) [inline Momento/Incubating/SimpleCacheClient.cs:16]: not a blocker for this PR, but I would probably expect this constructor to be private, and we would have a public one (or factory method) that handles the construction of the other client implicitly so the user doesn't have to.

Or perhaps you already have a builder somewhere that I haven't seen yet that takes care of this concern.

File: Momento/Incubating/SimpleCacheClientFactory.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+namespace MomentoSdk.Incubating;
+
+public class SimpleCacheClientFactory
+{
+    public static SimpleCacheClient Get(string authToken, uint defaultTtlSeconds, uint? dataClientOperationTimeoutMilliseconds = null)
```
Comment (cprice404) [inline Momento/Incubating/SimpleCacheClientFactory.cs:5]: Ah, here's the builder :)  ignore previous comment.

Is this a typical naming convention (`*Factory.Get`) for builders in CS?

**Review by cprice404 — APPROVED** [review]
Reviewed in more detail.  I like what you've done with the interface.  🚢 from me as soon as the more relevant stakeholders are happy :)

File: Momento/Incubating/SimpleCacheClientFactory.cs (hunk 15 above)
Comment (malandis) [reply Momento/Incubating/SimpleCacheClientFactory.cs:5]: > Is this a typical naming convention (*Factory.Get) for builders in CS?

I poked around and didn't see a consensus. I was learning toward `CreateClient` as opposed to `Get`. Thoughts?

**Review by kvcache — APPROVED** [review]

**Pull request merged**